### PR TITLE
store & use etags

### PR DESCRIPTION
**Pull request by yoshimo**
Signed-off-by: Yoshimo <yoshimo@users.noreply.github.com>

- **What does this PR aim to accomplish?:**

It tries to save bandwith by avoiding unneccessary downloads of files that haven't changed on the remote location since we last downloaded them.

- **How does this PR accomplish the above?:**
If the remote server supplies an etag header we store it for every list locally.
When we refresh a list we check if there is an etag stored for this entry and if we have an etag we add a "if-none-match: <etag>" header to the request.
The remote server compares our etag with its local database and if the conclusion is, that there hasn't been any change on the file, it will send us back a 304 not-modified status code.
We then use our local copy, avoid an unneccessary re-download and move on to the next entry on our list.

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

### CODE REVIEW

**Comment by yoshimo** [comment]
It seems to work fine most of the time. Github plays fine with the etags i send, but gitlab sends a 200 code and the same etag that i gave them. 
I must have missed something small.

**Comment by PromoFaux** [comment]
Hi, thanks for the submission.

> It tries to save bandwith by avoiding unneccessary downloads

I'll be honest, this is not a problem I've noticed personally, and wonder if this change adds unnessacery complexity (read: making it harder to maintain the change in future)

Do you have any stats you can provide to who how this would be useful?

**Comment by yoshimo** [comment]
> I'll be honest, this is not a problem I've noticed personally, and wonder if this change adds unnessacery complexity (read: making it harder to maintain the change in future)
> 
> Do you have any stats you can provide to who how this would be useful?

The idea is, to help people with slow connections and/or large block lists.
I am not sure how to measure the benefit exactly as a large part of the update process is spend on processing the downloaded files. An old Pi and a fast connection is probably the opposite of a good test caste.

**Comment by jfb-pihole** [comment]
> Number of gravity domains: 25577368 (12852714 unique domains)

This is a bit of an edge case.  25+ million domains is way outside of what almost any other user sees.

In any event, I suspect that in this case the vast majority of the time is in processing the domains (from either locally cached files or newly downloaded files), creating a new database and then swapping the existing gravity database for a new one.

The amount of time Pi-hole is offline is very short - just the time needed to swap the databases.  Everything prior to that is done in the background, with no impact on Pi-hole operation.

**Comment by MichaIng** [comment]
I'd probably put the additional output behind a debug flag (for testing or is nice) and assure that the Etag is really set (not empty) before storing it. Others a good idea.

**Comment by yoshimo** [comment]
> I'd probably put the additional output behind a debug flag (for testing it is nice) and assure that the Etag is really set (not empty) before storing it. 

I will redo the debug part once i am sure why adresses like

```raw.githubusercontent.com/chadmayfield/my-pihole-blocklist```
work but

```
gitlab.com/nobodysu/spyware-hosts/raw/master/lists/unsorted.txt
```
seems to ignore the etag.

The worst thing for now is that

```  [✗] Status: https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt (400)```
are caused with these changes and i haven't figured out what the remote server doesn't like

File: gravity.sh [hunk 1]
```diff
@@ -640,8 +640,27 @@ gravity_DownloadBlocklistFromUrl() {
...
+	  currentetag=$""
+  fi
+
+  echo -e "\n current Etag: " "$currentetag"
   # shellcheck disable=SC2086
-  httpCode=$(curl -s -L ${compression} ${cmd_ext} ${heisenbergCompensator} -w "%{http_code}" -A "${agent}" "${url}" -o "${patternBuffer}" 2> /dev/null)
+  httpCode=$(curl -s -L ${compression} ${cmd_ext} ${heisenbergCompensator} --trace "$id".log --header "If-None-Match: ${currentetag}" -w "%{http_code}" -A "${agent}" "${url}" -o "${patternBuffer}" 2> /dev/null)
+
+# shellcheck disable=SC2086
+  curl -sI "${url}" -L ${compression} ${cmd_ext} ${heisenbergCompensator} | tr -d "\r" | sed -e "/etag/!d" -e "s/^etag: //" | tr -d "\n" > "${piholeDir}/list.${id}.${domain}.etag"
```
Comment (MichaIng) [inline gravity.sh:656]: Allow ETag header with capital letters and use single quotes for literal strings:
```suggestion
  curl -sI "${url}" -L ${compression} ${cmd_ext} ${heisenbergCompensator} | tr -d '\r' | sed -e '/etag/I!d' -e 's/^etag: //I' | tr -d '\n' > "${piholeDir}/list.${id}.${domain}.etag"
```

Comment (yoshimo) [reply gravity.sh:656]: Still not foolproof :)
`https://urlhaus.abuse.ch/downloads/hostfile/` 
and 
`https://urlhaus.abuse.ch/downloads/text`

get the etag of

```
content-security-policy: default-src 'self' https://fonts.gstatic.com:443 data:; style-src 'self' 'unsafe-inline' https://www.gstatic.com:443 https://fonts.googleapis.com:443; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.gstatic.com:443 https://www.google.com/recaptcha/ https://www.googletagmanager.com:443; frame-src https://www.google.com/recaptcha/; img-src 'self' data: https://syndication.twitter.com:443; object-src 'none'"46ba9a-5e325b1fa769d-gzip"
```

this one is also funny
```
 [i] Target: https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/unity-all.txt
  [i] Status: Pending...
 current Etag:  W/"ee7bf2ecae84e0e39954ef52c279c56a"
 new Etag:       W/"ee7bf2ecae84e0e39954ef52c279c56a"
 statuscode:  200
  [✓] Status: Retrieval successful
  [i] Analyzed 143 domains
  [i] List stayed unchanged
```

**Comment by MichaIng** [comment]
Btw, see `heisenbergCompensator`: At least if the `Last-Modified` response header is set, the download will be skipped already. True for e.g. `https://urlhaus.abuse.ch/downloads/text`, but not for `https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/unity-all.txt`.

File: gravity.sh (hunk 1 above)
Comment (yoshimo) [reply gravity.sh:656]: yes, curl in a recent version would be nice but unfortunately pi hole runs on older versions to so i had to come up with the hack we are trying to make work here.

**Comment by yoshimo** [comment]
Urlhaus breaks parsing because of

```
content-security-policy: default-src 'self' https://fonts.gstatic.com:443 data:; style-src 'self' 'unsafe-inline' https://www.gs
tatic.com:443 https://fonts.googleapis.com:443; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.gstatic.com:443 https://www.google.com/recaptcha/ https://www.googletagmanager.com:443; frame-src https://www.google.com/recaptcha/; img-src 'self' data: https://syndication.twitter.com:443; object-src 'none'
```

```https://www.googlETAGmanager.com``` matches `etag `i guess

**Comment by MichaIng** [comment]
> `https://www.googlETAGmanager.com` matches `etag` i guess

`sed -n 's/^etag: //Ip'` should rule that out, as HEAD lines always begin with the header names like that. `curl` should assure line breaks always and only before the next header name.

**Comment by yoshimo** [comment]
> > `https://www.googlETAGmanager.com` matches `etag` i guess
> 
> `sed -n 's/^etag: //Ip'` should rule that out, as HEAD lines always begin with the header names like that. 

So i have run this version of the sed-command and i have run into the following bad cases

```  [i] Target: https://urlhaus.abuse.ch/downloads/text/
  [i] Status: Pending...
 current Etag:
 statuscode:  200
new Etag:      content-security-policy: default-src 'self' https://fonts.gstatic.com:443 data:; style-src 'self' 'unsafe-inline' https://www.gstatic.com:443 https://fonts.googleapis.com:443; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.gstatic.com:443 https://www.google.com/recaptcha/ https://www.googletagmanager.com:443; frame-src https://www.google.com/recaptcha/; img-src 'self' data: https://syndication.twitter.com:443; object-src 'none'```


```
  [i] Target: https://hosts.oisd.nl/
  [i] Status: Pending...
 current Etag:
 statuscode:  200
 new Etag:      access-control-expose-headers: ETag
"d2e037aa"
  [✓] Status: Retrieval successful
```

```
  [i] Target: https://urlhaus.abuse.ch/downloads/text/
  [i] Status: Pending...
 current Etag:
 statuscode:  200
 new Etag:      content-security-policy: default-src 'self' https://fonts.gstatic.com:443 data:; style-src 'self' 'unsafe-inline' https://www.gstatic.com:443 https://fonts.googleapis.com:443; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.gstatic.com:443 https://www.google.com/recaptcha/ https://www.googletagmanager.com:443; frame-src https://www.google.com/recaptcha/; img-src 'self' data: https://syndication.twitter.com:443; object-src 'none'
"3f42ae-5e4050005d3f6-gzip```

Gitlab still likes to send  a code 200 although i send the right etag , which is identical to the one i got back with the 200 code.
curl trace-ascii looks fine.

**Comment by yoshimo** [comment]
So i have a new thesis what is going on: Escaping of quotes.

```curl -sI --header "If-None-Match: W/\"53df85ca8b85e81646bbf90fd8114877\"" https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/qualcomm-mild.txt```
gives me a 304 not modified on the shell

```
httpCode=$(curl -s -L ${compression} ${cmd_ext} ${heisenbergCompensator} --header "If-None-Match: ${currentetag}" -w "%{http_code}" -A "${agent}" "${url}" -o "${patternBuffer}" 2> /dev/null)```
with the same etag in a file gives me 200 and the same etag back.

**Comment by rdwebdesign** [comment]
I'm not following the whole conversation here, but I think if your `$currenttag` variable uses single quotes it will work.

```
 curl -sI --header "If-None-Match: W/'53df85ca8b85e81646bbf90fd8114877'" https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/qualcomm-mild.txt
 ```

**Comment by dschaper** [comment]
The `etag` from the server contains a quote char, which is inherently stupid. 

Use single quotes around the complete header.

`curl -sI --header 'If-None-Match: W/"53df85ca8b85e81646bbf90fd8114877"' https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/qualcomm-mild.txt`

**Comment by dschaper** [comment]
And I'm happy to change the `curl` command in `gravity.sh` to get rid of the `heisenberg` and `patternbuffer` stuff that just confuses people now.

**Comment by yoshimo** [comment]
> Use single quotes around the complete header.
Single Quotes don't behave well with expanding the variable though.

**Comment by dschaper** [comment]
How so? Single quote around the `curl` command and double quote the variable expansion/substitution.

**Comment by dschaper** [comment]
`httpCode=$(curl -s -L ${compression} ${cmd_ext} ${heisenbergCompensator} --header 'If-None-Match: "${currentetag}"' -w "%{http_code}" -A "${agent}" "${url}" -o "${patternBuffer}" 2> /dev/null)`

**Comment by yoshimo** [comment]
> `httpCode=$(curl -s -L ${compression} ${cmd_ext} ${heisenbergCompensator} --header 'If-None-Match: "${currentetag}"' -w "%{http_code}" -A "${agent}" "${url}" -o "${patternBuffer}" 2> /dev/null)`

That will create:
```if-none-match: "${currentetag}"```

**Comment by rdwebdesign** [comment]
Can you please try:

```sh
httpCode=$(curl -s -L ${compression} ${cmd_ext} ${heisenbergCompensator} --header "'If-None-Match: ${currentetag}'" -w "%{http_code}" -A "${agent}" "${url}" -o "${patternBuffer}" 2> /dev/null)
```

**Comment by yoshimo** [comment]
`'if-none-match`  no, i think we need to skip the heisenberg compensator if we have a non-empty etag.

**Comment by dschaper** [comment]
> if-modified-since: Fri, 22 Jul 2022 17:17:08 GMT

Was the file modified in the few moments between that time and when you ran `curl`?

**Comment by yoshimo** [comment]
Some of these files in my list are changing often.
I think the current state works very well.
I only got a 200 statuscode 5 out of 161 requests where the file changed since the last time i collected data.
Both Gitlab and Github reply with  a 304 not modified if they only get one of the two possible headers.

**Comment by dschaper** [comment]
Two things:

One, can you give me the exact URLs for a 200 response and a 304 response with the actual values substituted in for the variables? I'd like to confirm if my suspicion about the modified time is correct. 

Two, please don't force push changes without saying something. It's really hard to check the PR locally if it is changing without anyone but you knowing it.

**Comment by yoshimo** [comment]
`curl -sI -H "Host: gitlab.com" -H "user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36" -H "accept: */*" -H "if-modified-since: Fri, 23 Jul 2021 17:31:38 GMT" -H "if-none-match: W/\"bfc13ed0e0799683836165c06b2615da\"" --compressed "https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/unsorted.txt"`

will give 200 and a small file that barely changes, Doesn't Matter if i use this or

`curl -sI -H "Host: gitlab.com" -H "user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36" -H "accept: */*" -H "if-modified-since: Fri, 22 Jul 2022 17:31:38 GMT" -H "if-none-match: W/\"bfc13ed0e0799683836165c06b2615da\"" --compressed "https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/unsorted.txt"`

To get a 304 nothing changed, you need to remove the `If-Modified-since`

`curl -sI -H "Host: gitlab.com" -H "user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36" -H "accept: */*"  -H "if-none-match: W/\"bfc13ed0e0799683836165c06b2615da\"" --compressed "https://gitlab.com/nobodysu/spyware-hosts/raw/master/lists/unsorted.txt"`

**Comment by DL6ER** [comment]
I have three comments after looking briefly at this:

1. We had one HTTP request (either download + code 200 or no download and some other response code) before. Now you are making an *independent* second HTTP requests just to get the ETag. Consider just making one request and using the `-D <filename>` option to store the received headers in a file from where you can read what you need without making a second request.
2. You are storing the extracted ETags in dedicated files, please don't do this. We store all details about the adlists (such as their address, when it was added, when it changed last, how many domains it has, ...) in `gravity.db` ([SQL schema](https://github.com/pi-hole/pi-hole/blob/aefbe1f5b6225b7476fe08264c62da84aaa61fa0/advanced/Templates/gravity.db.sql#L27-L39)). The ETag, being a simple header string, should be stored there as well in an extra column `ETag TEXT` (can be `NULL`).
3. Concerning the question about capitalization of `if-none-match` and if (and how) the value should be enquoted, I seem to recall that the RFC said something like "just record the exact header you received". So we should simply adhere to this and store the entire string, whether it is
   `if-none-match: "abcdef123456789abcdef"`, or
   `If-None-Match: W/"abcdef123456789abcd"`
   won't matter in this case. We just store the exact header this very server wants to see.

**Comment by yoshimo** [comment]
https://www.rfc-editor.org/rfc/rfc7230#section-3.2

``` 
Each header field consists of a case-insensitive field name followed
by a colon (":"), optional leading whitespace, the field value, and
optional trailing whitespace.
```
so that part is ok, it is not case-sensitive.

Haven't done database stuff , so that might take a bit off time to get it working.

**Comment by MichaIng** [comment]
> Consider just making one request and using the `-D <filename>` option to store the received headers in a file from where you can read what you need without making a second request.

Seconded, same idea here. Could be also stored in a variable only instead of a file.
